yarn-project/prover-client/src/proving_broker/broker_prover_facade.ts

```diff
@@ -53,77 +73,282 @@ export class BrokerCircuitProverFacade implements ServerCircuitProver {
+        while (ids.length > 0) {
+          const slice = ids.splice(0, SNAPSHOT_SYNC_CHECK_MAX_REQUEST_SIZE);
+          const completed = await this.broker.getCompletedJobs(slice);
+          completed.forEach(id => allCompleted.add(id));
+          ++numRequests;
+        }
+        if (numRequests === 0) {
+          const final = await this.broker.getCompletedJobs([]);
+          final.forEach(id => allCompleted.add(id));
+        }
+      } catch (err) {
+        this.log.error(`Error thrown when requesting completed job notifications from the broker`, err);
+      }
+      return allCompleted;
+    };
+
+    const snapshotSyncIds = [];
+    const currentTime = Date.now();
+    const secondsSinceLastSnapshotSync = currentTime - this.timeOfLastSnapshotSync;
+    if (secondsSinceLastSnapshotSync > SNAPSHOT_SYNC_INTERVAL_MS) {
+      this.timeOfLastSnapshotSync = currentTime;
+      snapshotSyncIds.push(...this.jobs.keys());
+      this.log.trace(`Performing full snapshot sync of completed jobs with ${snapshotSyncIds.length} job(s)`);
+    } else {
+      this.log.trace(`Performing incremental sync of completed jobs`);
+    }
+
+    // Now request the notifications from the broker
+    const snapshotIdsLength = snapshotSyncIds.length;
+    const completedJobs = await getAllCompletedJobs(snapshotSyncIds);
+
+    // We now have an additional set of completed job notifications to add to our cached set giving us the full set of jobs that we have been told are ready
+    // We filter this list to what we actually need, in case for any reason it is different and store in our cache
+    const allJobsReady = [...completedJobs, ...this.jobsToRetrieve];
+    this.jobsToRetrieve = new Set(allJobsReady.filter(id => this.jobs.has(id)));
+
+    if (completedJobs.size > 0) {
+      this.log.verbose(
+        `Check for job completion notifications returned ${completedJobs.size} job(s), snapshot ids length: ${snapshotIdsLength}, num outstanding jobs: ${this.jobs.size}, total jobs ready: ${this.jobsToRetrieve.size}`,
+      );
+    } else {
+      this.log.trace(
+        `Check for job completion notifications returned 0 jobs, snapshot ids length: ${snapshotIdsLength}, num outstanding jobs: ${this.jobs.size}, total jobs ready: ${this.jobsToRetrieve.size}`,
+      );
+    }
+  }
+
+  private async retrieveJobsThatShouldBeReady() {
+    const convertJobResult = async <T extends ProvingRequestType>(
+      result: ProvingJobStatus,
+      jobType: ProvingRequestType,
+    ): Promise<{
+      success: boolean;
+      reason?: string;
+      result?: ProvingJobResultsMap[T];
+    }> => {
       if (result.status === 'fulfilled') {
         const output = await this.proofStore.getProofOutput(result.value);
-        if (output.type === type) {
-          return output.result as ProvingJobResultsMap[T];
+        if (output.type === jobType) {
+          return { result: output.result as ProvingJobResultsMap[T], success: true, reason: '' };
```

Comment:
```suggestion
          return { result: output.result as ProvingJobResultsMap[T], success: true };
```
Nit: let's not use empty strings instead of null/undefined.

Reply:
Done